### PR TITLE
FastAPI services: remove limitation of 1 connection per request

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
- [x] director-v2 now acquires connection when executing transactions
- [x] director-v2 max number of connections set back to default of 10 
- [x] catalog now acquires connection when executing transactions
- [x] api-server

Bonus:
- [x] webserver/list_projects gets the study states AFTER splicing (e.g. if call with *count=1* only ask for 1 state not all of them and then splice)
- [x] fix missing docker logs in e2e CI testing

### CODE REVIEW

**Review by pcrespov — APPROVED** [review]
It is a good first iteration but I would like to make it even more compact,

e.g. by defining ``acquire()`` in the base class and log number of connections stats for that service as we used to do before

File: ci/github/system-testing/e2e.bash [hunk 1]
```diff
@@ -50,16 +48,16 @@ dump_docker_logs() {
   # get docker logs
   # NOTE: Timeout avoids issue with dumping logs that hang!
   mkdir --parents simcore_logs
-  (timeout 30 docker service logs --timestamps --tail=300 --details ${SWARM_STACK_NAME}_webserver >simcore_logs/webserver.log 2>&1) || true
+  (timeout 30 docker service logs --timestamps --tail=300 --details "${SWARM_STACK_NAME:-test}"_webserver >simcore_logs/webserver.log 2>&1) || true
```
Comment (GitHK) [inline ci/github/system-testing/e2e.bash:51]: 👍

File: services/catalog/src/simcore_service_catalog/db/repositories/dags.py [hunk 2]
```diff
@@ -13,48 +13,55 @@
...
-        row: RowProxy = await (await self.connection.execute(stmt)).first()
-        if row:
-            return DAGAtDB(**row)
-        return None
+        async with self.db_engine.acquire() as conn:
+            row: RowProxy = await (
+                await conn.execute(dags.select().where(dags.c.id == dag_id))
+            ).first()
+            if row:
+                return DAGAtDB(**row)
```
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/db/repositories/dags.py:28]: could you reduce indentation by 1 level here, to guaranty an early release of the connection?

File: services/catalog/src/simcore_service_catalog/db/repositories/groups.py [hunk 3]
```diff
@@ -13,38 +13,42 @@
...
-            )
-        ).first()
-        return GroupAtDB(**row)
+        async with self.db_engine.acquire() as conn:
+            row: RowProxy = await (
+                await conn.execute(
+                    sa.select([groups]).where(groups.c.type == GroupType.EVERYONE)
+                )
+            ).first()
+            return GroupAtDB(**row)
```
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/db/repositories/groups.py:34]: also reduce indentation here

File: services/catalog/src/simcore_service_catalog/db/repositories/services.py [hunk 4]
```diff
@@ -86,49 +85,52 @@ async def get_service(
                     )
                 )
             )
-        row: RowProxy = await (await self.connection.execute(query)).first()
-        if row:
+        async with self.db_engine.acquire() as conn:
+            row: RowProxy = await (await conn.execute(query)).first()
             return ServiceMetaDataAtDB(**row)
```
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/db/repositories/services.py:90]: also reduce indentation here

File: services/catalog/src/simcore_service_catalog/db/repositories/services.py [hunk 5]
```diff
@@ -86,49 +85,52 @@ async def get_service(
...
                 )
-                .values(**patched_service.dict(by_alias=True, exclude_unset=True))
-                .returning(literal_column("*"))
-            )
-        ).first()
-        updated_service = ServiceMetaDataAtDB(**row)
-        return updated_service
+            ).first()
+            updated_service = ServiceMetaDataAtDB(**row)
+            return updated_service
```
Comment (GitHK) [inline services/catalog/src/simcore_service_catalog/db/repositories/services.py:133]: reduce indentation here as well

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_pipelines.py [hunk 6]
```diff
@@ -19,15 +19,16 @@
...
             )
-        )
-        row: RowProxy = await result.fetchone()
-        if not row:
-            raise PipelineNotFoundError(str(project_id))
-        return CompPipelineAtDB.from_orm(row)
+            row: RowProxy = await result.fetchone()
+            if not row:
+                raise PipelineNotFoundError(str(project_id))
+            return CompPipelineAtDB.from_orm(row)
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_pipelines.py:31]: reduce indentation of this block, if possible

File: services/director-v2/src/simcore_service_director_v2/modules/db/repositories/projects.py [hunk 7]
```diff
@@ -13,11 +13,12 @@
...
-        return ProjectAtDB.from_orm(row)
+        async with self.db_engine.acquire() as conn:
+            row: RowProxy = await (
+                await conn.execute(
+                    sa.select([projects]).where(projects.c.uuid == str(project_id))
+                )
+            ).first()
+            if not row:
+                raise ProjectNotFoundError(project_id)
+            return ProjectAtDB.from_orm(row)
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/db/repositories/projects.py:24]: also reduce indentation if applies

**Review by GitHK — APPROVED** [review]
Looks nice, added a few requests to reduce indentation (early connection releases). I am not sure they apply in all cases.

**Comment by sanderegg** [comment]
> It is a good first iteration but I would like to make it even more compact,
> 
> e.g. by defining `acquire()` in the base class and log number of connections stats for that service as we used to do before

1 question there. I was thinking about this. but do I get it right, that the only thing this would bring is the logging of the number of connection right? the rest would still work with a context manager a la:
```python
async with acquire() as _:
  # use the connection
```
correct?

File: services/catalog/src/simcore_service_catalog/db/repositories/dags.py (hunk 2 above)
Comment (sanderegg) [reply services/catalog/src/simcore_service_catalog/db/repositories/dags.py:28]: actually learned something this morning about scope of with, if and for in python... very weird though. thanks

Comment (sanderegg) [reply services/catalog/src/simcore_service_catalog/db/repositories/dags.py:28]: but now. what if row is None.. this will fail right?

Comment (sanderegg) [reply services/catalog/src/simcore_service_catalog/db/repositories/dags.py:28]: ah no. I keep the whole block all good. thanks

**Comment by pcrespov** [comment]
> > It is a good first iteration but I would like to make it even more compact,
> > e.g. by defining `acquire()` in the base class and log number of connections stats for that service as we used to do before
> 
> 1 question there. I was thinking about this. but do I get it right, that the only thing this would bring is the logging of the number of connection right? the rest would still work with a context manager a la:
> 
> ```python
> async with acquire() as _:
>   # use the connection
> ```
> 
> correct?

yes

**Pull request merged**